File: py/2.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        head = ListNode()
        last = head
        carry = 0
        while l1 != None and l2 != None:
            result = l1.val + l2.val + carry
            carry = result // 10
            current = ListNode(result % 10)
            last.next = current
            last = current
            l1 = l1.next
            l2 = l2.next
        while l1 != None:
            result = l1.val + carry
            carry = result // 10
            current = ListNode(result % 10)
            last.next = current
            last = current
            l1 = l1.next
        while l2 != None:
            result = l2.val + carry
            carry = result // 10
            current = ListNode(result % 10)
            last.next = current
            last = current
            l2 = l2.next
        if carry > 0:
            current = ListNode(1)
            last.next = current
        return head.next

File: cpp/2.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode *head = new ListNode(), *prev = head;
        int carry = 0;
        while (l1 && l2) {
            int sum = l1->val + l2->val + carry;
            carry = sum / 10;
            sum %= 10;
            prev->next = new ListNode(sum);
            l1 = l1->next;
            l2 = l2->next;
            prev = prev->next;
        }
        while (l1) {
            int sum = l1->val + carry;
            carry = sum / 10;
            sum %= 10;
            prev->next = new ListNode(sum);
            l1 = l1->next;
            prev = prev->next;
        }
        while (l2) {
            int sum = l2->val + carry;
            carry = sum / 10;
            sum %= 10;
            prev->next = new ListNode(sum);
            l2 = l2->next;
            prev = prev->next;
        }
        if (carry)
            prev->next = new ListNode(carry);
        return head->next;
    }
};